File: app/schemas/agent.py
"""app/schemas/agent.py — 交易员市场 schema"""
from datetime import datetime, date
from pydantic import BaseModel, Field


class AgentTraderResponse(BaseModel):
    id: str
    code_name: str
    tag: str
    avatar_url: str | None = None
    description: str
    strategy_detail: str | None = None
    masters: str
    hire_price_points: int
    profit_share_pct: float
    annual_return: float | None = None
    max_drawdown: float | None = None
    sharpe_ratio: float | None = None
    win_rate: float | None = None
    total_trades: int | None = None
    radar_scores: dict | None = None
    salary_curve: list | None = None

    class Config:
        from_attributes = True


class AgentTraderDetail(AgentTraderResponse):
    is_hired: bool = False
    management_mode: str | None = None
    hired_at: datetime | None = None
    current_pnl: float | None = None
    recent_performances: list["AgentPerformanceResponse"] = []


class AgentPerformanceResponse(BaseModel):
    period: str
    period_end: date
    return_pct: float
    benchmark_return_pct: float | None = None
    alpha: float | None = None
    max_drawdown: float | None = None
    sharpe_ratio: float | None = None
    win_rate: float | None = None

    class Config:
        from_attributes = True


class AgentPerformanceDetailResponse(BaseModel):
    agent_id: str
    performance_metrics: AgentPerformanceResponse | None = None
    salary_curve: list[dict] = []
    recent_performances: list[AgentPerformanceResponse] = []


class AgentMarketListResponse(BaseModel):
    items: list[AgentTraderResponse]
    total: int


class HireAgentRequest(BaseModel):
    management_mode: str = Field(default="advisory", pattern=r"^(advisory|full_managed)$")


class HireAgentResponse(BaseModel):
    user_agent_id: int
    agent_id: str
    points_spent: int
    balance_after: int
    management_mode: str
    expires_at: datetime | None = None
    message: str = "雇佣成功"


class UserAgentResponse(BaseModel):
    id: int
    agent_id: str
    agent: AgentTraderResponse
    status: str
    management_mode: str
    allocated_capital: float | None = None
    current_pnl: float | None = None
    hired_at: datetime
    expires_at: datetime | None = None
    config_source: str = "default"

    class Config:
        from_attributes = True


class UpdateManagementModeRequest(BaseModel):
    management_mode: str = Field(pattern=r"^(advisory|full_managed)$")


# ── 控制台 Schemas ──

class ConsoleOverviewResponse(BaseModel):
    hire_id: int
    trader_name: str
    trader_tag: str = ""
    management_mode: str
    status: str
    total_assets: float = 0
    unrealized_pnl: float = 0
    today_signals: int = 0
    pending_signals: int = 0
    position_count: int = 0


class ConsoleSignalResponse(BaseModel):
    id: int
    hire_id: int
    trader_id: str
    symbol: str
    symbol_name: str
    market: str
    action: str
    price: float
    quantity: int
    confidence: int
    reasoning: str | None = None
    exec_status: str
    created_at: datetime
    updated_at: datetime | None = None

    class Config:
        from_attributes = True


class SignalConfirmRequest(BaseModel):
    quantity: int | None = None


class ConsolePortfolioResponse(BaseModel):
    id: int
    hire_id: int
    symbol: str
    symbol_name: str
    quantity: int
    avg_cost: float
    current_price: float | None = None
    market_value: float | None = None
    unrealized_pnl: float | None = None

    class Config:
        from_attributes = True


class ConsoleTradeResponse(BaseModel):
    id: int
    symbol: str
    symbol_name: str
    action: str
    price: float
    quantity: int
    confidence: int
    reasoning: str | None = None
    exec_status: str
    executed_at: datetime | None = None


class EquityCurvePoint(BaseModel):
    date: str
    equity: float
    daily_pnl: float = 0


AgentTraderDetail.model_rebuild()


# ── Agent 配置 Schemas ──

class AgentConfigRequest(BaseModel):
    """配置请求（所有字段均可选，只传要修改的）"""
    markets: list[str] | None = None
    trade_hours_pref: str | None = None
    dormant_off_hours: bool | None = None
    sectors: list[str] | None = None
    market_cap_min: float | None = None
    market_cap_max: float | None = None
    min_avg_amount: float | None = None
    exclude_st: bool | None = None
    exclude_new_listing_days: int | None = None
    exclude_limit_near: bool | None = None
    allocated_capital: float | None = None
    max_order_pct: float | None = None
    max_position_pct: float | None = None
    max_position_count: int | None = None
    trading_style: str | None = None
    signal_interval_min: int | None = None
    daily_max_signals: int | None = None
    daily_max_executions: int | None = None
    daily_max_amount_pct: float | None = None
    loss_stop_pct: float | None = None
    loss_stop_amount: float | None = None
    risk_trigger_action: str | None = None
    risk_enabled: bool | None = None
    t1_enabled: bool | None = None
    auto_exec_confidence: int | None = None
    max_auto_exec_per_round: int | None = None
    analyze_ticker_limit: int | None = None
    auto_exec_enabled: bool | None = None
    notify_channels: list[str] | None = None


class AgentConfigResponse(BaseModel):
    """配置响应"""
    hire_id: int
    markets: list[str]
    trade_hours_pref: str
    dormant_off_hours: bool
    sectors: list[str]
    market_cap_min: float | None = None
    market_cap_max: float | None = None
    min_avg_amount: float
    exclude_st: bool
    exclude_new_listing_days: int
    exclude_limit_near: bool
    allocated_capital: float
    max_order_pct: float
    max_position_pct: float
    max_position_count: int
    trading_style: str
    signal_interval_min: int
    daily_max_signals: int
    daily_max_executions: int
    daily_max_amount_pct: float
    loss_stop_pct: float
    loss_stop_amount: float | None = None
    risk_trigger_action: str
    risk_enabled: bool
    t1_enabled: bool
    auto_exec_confidence: int
    max_auto_exec_per_round: int
    analyze_ticker_limit: int = 10
    auto_exec_enabled: bool = True
    notify_channels: list[str]
    config_source: str = "default"
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ActivateAgentResponse(BaseModel):
    """启用交易员响应"""
    user_agent_id: int
    status: str
    message: str


class DeactivateAgentResponse(BaseModel):
    """停用交易员响应"""
    user_agent_id: int
    status: str
    message: str

File: src/api/agent.ts
/**
 * 交易员市场 API
 */
import { request } from '@/utils/request'

export interface AgentTrader {
  id: string
  code_name: string
  tag: string
  avatar_url: string | null
  description: string
  strategy_detail: string | null
  masters: string
  hire_price_points: number
  profit_share_pct: number
  annual_return: number | null
  max_drawdown: number | null
  sharpe_ratio: number | null
  win_rate: number | null
  total_trades: number | null
}

export interface AgentPerformance {
  period: string
  period_end: string
  return_pct: number
  benchmark_return_pct: number | null
  alpha: number | null
  max_drawdown: number | null
  sharpe_ratio: number | null
  win_rate: number | null
}

export interface AgentTraderDetail extends AgentTrader {
  is_hired: boolean
  management_mode: string | null
  hired_at: string | null
  current_pnl: number | null
  recent_performances: AgentPerformance[]
}

export interface AgentMarketList {
  items: AgentTrader[]
  total: number
}

export interface HireAgentResponse {
  user_agent_id: number
  agent_id: string
  points_spent: number
  balance_after: number
  management_mode: string
  expires_at: string | null
  message: string
}

export interface UserAgent {
  id: number
  agent_id: string
  agent: AgentTrader
  status: string
  management_mode: string
  allocated_capital: number | null
  current_pnl: number | null
  hired_at: string
  expires_at: string | null
  config_source?: string
}

/** 获取交易员市场列表 */
export function getAgentMarket(): Promise<AgentMarketList> {
  return request<AgentMarketList>('/agent/market')
}

/** 获取交易员详情 */
export function getAgentDetail(agentId: string): Promise<AgentTraderDetail> {
  return request<AgentTraderDetail>('/agent/market/' + agentId)
}

/** 雇佣交易员 */
export function hireAgent(agentId: string, managementMode: string): Promise<HireAgentResponse> {
  return request<HireAgentResponse>('/agent/market/' + agentId + '/hire', {
    method: 'POST',
    data: { management_mode: managementMode },
  })
}

/** 获取我的交易员列表 */
export function getMyAgents(): Promise<UserAgent[]> {
  return request<UserAgent[]>('/agent/my-agents')
}

/** 切换管理模式 */
export function updateManagementMode(userAgentId: number, managementMode: string): Promise<any> {
  return request('/agent/my-agents/' + userAgentId + '/mode', {
    method: 'PATCH',
    data: { management_mode: managementMode },
  })
}

/** 解雇交易员 */
export function dismissAgent(userAgentId: number): Promise<any> {
  return request('/agent/my-agents/' + userAgentId, { method: 'DELETE' })
}

// ── 控制台 API ──

export interface ConsoleOverview {
  hire_id: number
  trader_name: string
  trader_tag: string
  management_mode: string
  status: string
  total_assets: number
  unrealized_pnl: number
  today_signals: number
  pending_signals: number
  position_count: number
}

export interface ConsoleSignal {
  id: number
  hire_id: number
  trader_id: string
  symbol: string
  symbol_name: string
  action: string
  price: number
  quantity: number
  confidence: number
  reasoning: string | null
  exec_status: string
  created_at: string
  updated_at: string | null
}

export interface ConsolePortfolio {
  id: number
  hire_id: number
  symbol: string
  symbol_name: string
  quantity: number
  avg_cost: number
  current_price: number | null
  market_value: number | null
  unrealized_pnl: number | null
}

export interface ConsoleTrade {
  id: number
  symbol: string
  symbol_name: string
  action: string
  price: number
  quantity: number
  confidence: number
  reasoning: string | null
  exec_status: string
  executed_at: string | null
}

export interface EquityCurvePoint {
  date: string
  equity: number
  daily_pnl: number
}

/** 获取控制台概览 */
export function getConsoleOverview(hireId: number): Promise<ConsoleOverview> {
  return request<ConsoleOverview>('/agent-console/' + hireId + '/overview')
}

/** 获取信号列表 */
export interface SignalFilters {
  status?: string
  symbol?: string
  date_from?: string
  date_to?: string
}

export function getSignals(hireId: number, filters?: SignalFilters): Promise<ConsoleSignal[]> {
  const params = new URLSearchParams()
  if (filters?.status) params.set('status', filters.status)
  if (filters?.symbol) params.set('symbol', filters.symbol)
  if (filters?.date_from) params.set('date_from', filters.date_from)
  if (filters?.date_to) params.set('date_to', filters.date_to)
  const qs = params.toString()
  return request<ConsoleSignal[]>('/agent-console/' + hireId + '/signals' + (qs ? '?' + qs : ''))
}

/** 确认信号执行 */
export function confirmSignal(signalId: number, quantity?: number): Promise<any> {
  return request('/agent-console/signals/' + signalId + '/confirm', {
    method: 'POST',
    data: quantity ? { quantity } : {},
  })
}

/** 忽略信号 */
export function ignoreSignal(signalId: number): Promise<any> {
  return request('/agent-console/signals/' + signalId + '/ignore', { method: 'POST' })
}

/** 获取交易员持仓 */
export function getAgentPortfolio(hireId: number): Promise<ConsolePortfolio[]> {
  return request<ConsolePortfolio[]>('/agent-console/' + hireId + '/portfolio')
}

/** 获取交易日志 */
export function getAgentTrades(hireId: number): Promise<ConsoleTrade[]> {
  return request<ConsoleTrade[]>('/agent-console/' + hireId + '/trades')
}

/** 获取权益曲线 */
export function getEquityCurve(hireId: number): Promise<EquityCurvePoint[]> {
  return request<EquityCurvePoint[]>('/agent-console/' + hireId + '/equity-curve')
}

// ── 实时交易大屏看板 API ──

export interface LiveBoardAgentStatus {
  hire_id: number
  trader_id: string
  trader_name: string
  trader_tag: string
  status: string
  runtime_status: string
  runtime_label: string
  management_mode: string
  current_pnl: number
  allocated_capital: number | null
  last_action: string | null
  last_action_at: string | null
}

export interface LiveBoardTrade {
  id: number
  hire_id: number
  trader_id: string
  trader_name: string
  symbol: string
  symbol_name: string
  market: string
  action: string
  price: number
  quantity: number
  confidence: number
  reasoning: string | null
  exec_status: string
  created_at: string
}

export interface LiveBoardResponse {
  agents: LiveBoardAgentStatus[]
  trades: LiveBoardTrade[]
  scheduler_running: boolean
  market_state: string
}

/** 获取实时交易大屏看板 */
export function getLiveBoard(): Promise<LiveBoardResponse> {
  return request<LiveBoardResponse>('/agent-console/live-board')
}

/** 获取单条信号详情（含决策理由） */
export function getSignalDetail(signalId: number): Promise<ConsoleSignal> {
  return request<ConsoleSignal>('/agent-console/signals/' + signalId)
}

/** 切换管理模式 */
export function switchMode(hireId: number, mode: string): Promise<any> {
  return request('/agent/my-agents/' + hireId + '/mode', {
    method: 'PATCH',
    data: { management_mode: mode },
  })
}

/** 暂停交易员 */
export function pauseAgent(hireId: number): Promise<any> {
  return request('/agent/my-agents/' + hireId + '/pause', { method: 'POST' })
}

/** 恢复交易员 */
export function resumeAgent(hireId: number): Promise<any> {
  return request('/agent/my-agents/' + hireId + '/resume', { method: 'POST' })
}

/** 终止雇佣 */
export function terminateAgent(hireId: number): Promise<any> {
  return request('/agent/my-agents/' + hireId + '/terminate', { method: 'POST' })
}

/** 手动生成信号 */
export function generateSignals(hireId: number): Promise<ConsoleSignal[]> {
  return request<ConsoleSignal[]>('/agent-console/' + hireId + '/generate-signals', { method: 'POST' })
}

// ── 交易员配置 API ──

export interface AgentConfig {
  hire_id?: number
  markets: string[]
  allocated_capital: number
  max_position_pct: number | null
  max_position_count: number | null
  loss_stop_pct: number | null
  loss_stop_amount: number | null
  t1_enabled: boolean
  auto_exec_confidence: number
  max_auto_exec_per_round: number | null
  signal_interval_min: number | null
  trading_style: string
}

/** 获取交易员配置 */
export function getAgentConfig(hireId: number): Promise<AgentConfig> {
  return request<AgentConfig>('/agent/my-agents/' + hireId + '/config')
}

/** 更新交易员配置 */
export function updateAgentConfig(hireId: number, data: Partial<AgentConfig>): Promise<AgentConfig> {
  return request<AgentConfig>('/agent/my-agents/' + hireId + '/config', {
    method: 'PATCH',
    data,
  })
}

/** 启用交易员 */
export function activateAgent(hireId: number): Promise<any> {
  return request('/agent/my-agents/' + hireId + '/activate', { method: 'POST' })
}
